Tuist run command

itsmathou opened the pull request:
Resolves https://github.com/tuist/tuist/issues/1650

### Short description 📝

> This is a draft with first ideas for the Run command. I was wondering if I could have some guidance if I'm in the right direction?

### Solution 📦

> Describe the solution you came up with and the reasons that led you to that solution. If you thought about other solutions don't forget about mentioning them.

### Implementation 👩‍💻👨‍💻

> Detail in a checklist the steps that you took to implement the PR.

- [ ] Finding a way to use the date suggested by Pedro
```
Note that we are passing after to make sure we are only taking apps that have been generated after we started compilation. That prevents taking apps that were previously compiled in the derived data directory.
```
- [ ] Make `findDevice` less... dirty

pepibumur commented on Sources/TuistAutomation/Simulator/SimulatorController.swift:
```diff
@@ -100,4 +106,12 @@ final class SimulatorController: SimulatorControlling {
+        logger.log(level: .notice, "Booting \(simulator.description)", metadata: .section)
+        _ = try System.shared.observable(["/usr/bin/xcrun", "simctl", "boot", "\(simulator.udid)"])
+            .mapToString()
+            .toBlocking()
```
@itsmathou what about returning a `Completble` from this method instead of making it synchronous? That way we can easily combine multiple asynchronous actions for managing the simulator, and run them in parallel to the compilation of the app.

pepibumur commented on Sources/TuistKit/Commands/RunCommand.swift:
```diff
@@ -0,0 +1,25 @@
+    @Argument(
+        help: "The scheme to be guild. By default it builds all the buildable schemes of the project in the current directory"
+    )
+    var schemeName: String
```
If I'm not wrong, I think the `@Argument` property wrapper will infer the following argument names: `-s` & `--scheme-name`. Although the first one I think is right, I think the second one could be shorter and more aligned with `xcodebuild`: `--scheme`. You can achieve that by either renaming `schemeName` to `scheme`, or passing the `name` argument to the initialization of the property wrapper:

```swift
@Argument(name: [.customShort("s"), .customLong("scheme")])
```

pepibumur commented on Sources/TuistKit/Commands/RunCommand.swift:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    @Argument(
+        help: "The scheme to be guild. By default it builds all the buildable schemes of the project in the current directory"
```
Typo: `guild`. I'd say something along the lines of:

```
The scheme to run.
```

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+        case .simulatorNotFound:
+            return "Could't find simulator"
+        case let .targetNotFound(forSchemeName):
+            return "Couldn't find \(forSchemeName)"
```
I'd extend it to include the list of schemes. That makes it easier for the user and for us to debug the issue:

```
Couldn't find scheme '\(schemeName)' in the project. The list of buildable schemes are: ....`
```

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+        case let .appNotFound(targetName):
+            return "Couldn't find any apps built for \(targetName)"
+        case .simulatorNotFound:
+            return "Could't find simulator"
```
An error like this would leave the user without knowing what to do. What about giving the user some tips on what to do in this case? What could be the reason for a simulator not being found?

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+}
+
+final class RunService {
+    let simulatorController: SimulatorControlling
```
Nitpick: I'd remove the empty lines between the definition of the instance variables.

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+    private func buildScheme(_ schemeName: String) throws {
+        let graph = try projectGenerator.load(path: FileHandler.shared.currentPath)
+        guard let scheme = buildGraphInspector.buildableSchemes(graph: graph).first(where: { $0.name == schemeName }) else {
+            throw RunServiceError.appNotFound(targetName: "")
```
Should the error here be `RunServiceError.schemeNotFound(scheme: schemeName, availableSchemes: schemes)`?

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    private func findApp(for target: Target) throws -> AbsolutePath {
+        let appsLocation = FileHandler.shared.locateDirectory("../Library/Developer/Xcode/DerivedData/",
```
I'd extend [XcodeController](https://github.com/tuist/tuist/blob/master/Sources/TuistSupport/Xcode/XcodeController.swift) with a method to get the derive data directory:

```
func derivedDataPath() throws -> AbsolutePath
```

Then, we could follow the logic described in [this blog post](https://pewpewthespells.com/blog/xcode_build_locations.html):
1. If custom location is defined, take the custom location.
2. Otherwise, read the default value `~/Library/Developer/Xcode/DerivedData`.

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+                                                              traversingFrom: FileHandler.shared.currentPath)?
+            .glob("\(target.name)-*")
+            .first?
+            .glob("Build/Products/Debug-iphonesimulator/*.app")
```
With `Debug-iphonesimulator` we are assuming that the target is for iOS, which might not be true. It's worth getting the right path based on the target's platform. I think you can use [this method](https://github.com/tuist/tuist/blob/master/Sources/TuistCore/Models/Platform.swift#L75) to get the information that you need. 

Also, I'd introduce a check that exits the execution if the target is for a platform other than iOS. We can tell the user that we only support iOS at the moment.

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+    private func findApp(for target: Target) throws -> AbsolutePath {
+        let appsLocation = FileHandler.shared.locateDirectory("../Library/Developer/Xcode/DerivedData/",
+                                                              traversingFrom: FileHandler.shared.currentPath)?
+            .glob("\(target.name)-*")
```
You can put the 2 globs together:

```
"\(target.name)-*/Build/Products/Debug-\(target.platform.xcodeSimulatorSDK)/*.app"
```

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+
+    private func installAppBuilt(for target: Target) throws {
+        logger.log(level: .notice, "Installing \(target.name)", metadata: .section)
+        _ = try System.shared.observable(["/usr/bin/xcrun", "simctl", "install", "booted", "\(appPath!)"])
```
This is a good candidate to be moved into the `SimulatorController` utility:

```swift
func installApp(device: Device, appPath: AbsolutePath) -> Completable
```

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+            .last()
+    }
+
+    private func launchAppOnSimulator(for target: Target) throws {
```
This is another candidate to be moved into the `SimulatorController` utility:

```swift
func launchApp(device: Device, bundleId: String -> Completable
```

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+        try simulatorController.devices()
+            .toBlocking()
+            .first()?
+            .filter { $0.description.contains("iPhone 11") }
```
I'd pick the first compatible simulator without checking the model.

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+    // Sort of wondering if it belongs here or in the SimulatorController class
+    // At some point we can imagine we'll have some arguments to determine which
+    // simulator to pick (device, OS version, etc)
+    private func findDevice() throws -> SimulatorDevice? {
```
I'd move this one into `SimulatorController` and return a `Single<SimulatorDevice>`

```swift
func findAvailableDevice(platform: Platform) -> Single<SimulatorDevice>
```

If a device can't be found we can raise an error instead of returning `nil`.

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+            .glob("Build/Products/Debug-iphonesimulator/*.app")
+
+        guard let locations = appsLocation else { throw RunServiceError.appNotFound(targetName: target.name) }
+        if locations.count == 0 {
```
You can do `locations.empty`

pepibumur commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -0,0 +1,142 @@
+        }
+
+        appPath = try findApp(for: buildableTarget)
+        try launchApp(for: buildableTarget)
```
Since you have the build process, which you can model with an `Observable`, and the simulator logic, which can also be wrapped inside an `Observable`, I think there's an opportunity here to do things in parallel.

1. You trigger the build of the app and the launch of the simulator in parallel.
2. If the build finishes successfully, then you install and launch the app.

I think you can model the logic with 3 observables:

1. buildAppObservable
2. setupDeviceObservable
3. installAndLaunchAppObservable.

1 and 2 can run in parallel, while 3 has to wait until the first 2 finish successfully.

pepibumur reviewed (commented):
Great job @itsmathou 👏. The PR is in a good direction.

pepibumur commented:
👋  @itsmathou
I'm doing a follow-up on this PR. Do you have plans to move it forward? Otherwise I can take it on and finalize it.

itsmathou commented:
> 👋 @itsmathou
> I'm doing a follow-up on this PR. Do you have plans to move it forward? Otherwise I can take it on and finalize it.

Hey Pable, I addressed some of the comments and struggled with others. I'll message you to talk further about them :)

fortmarek commented:
👋 @itsmathou I have added a new method to `SimulatorController` to find available device based on some parameters in [this](https://github.com/tuist/tuist/pull/1935) PR. I am not sure if it might of use to you but just passing it on to be sure 🙂
